common/src/main/java/com/wynntils/features/map/WorldMarkersFeature.java

```diff
@@ -0,0 +1,535 @@
+        if (isInBound(position, window)) return null;
+        Vec3 centerPoint = new Vec3(window.getGuiScaledWidth() / 2f, window.getGuiScaledHeight() / 2f, 0);
+
+        float pointerScaleCorrection = (float) Texture.POINTER.height() / 2 * (1 - scale.get());
+
+        // minecraft's origin point is top left corner
+        // so positive Y is at the screen bottom
+        float minX = (float) -(centerPoint.x - horizontalBoundingDistance.get() + pointerScaleCorrection);
+        float maxX = (float) centerPoint.x - horizontalBoundingDistance.get() + pointerScaleCorrection;
+        float minY = (float) -(centerPoint.y - topBoundingDistance.get() + pointerScaleCorrection);
+        float maxY = (float) centerPoint.y - bottomBoundingDistance.get() + pointerScaleCorrection;
+
+        // drag the origin point to center since indicator's screenspace position / rotation is in relation to it
+        Vec3 centerRelativePosition = position.subtract(centerPoint);
+
+        // invert xy axis if target is behind camera
+        if (centerRelativePosition.z > 1) {
+            centerRelativePosition = centerRelativePosition.multiply(-1, -1, 1);
+        }
+
+        // since center point is now the origin point, atan2 is used to get the angle, and angle is used to get the
+        // line's slope
+        double angle = StrictMath.atan2(centerRelativePosition.y, centerRelativePosition.x);
+        double m = StrictMath.tan(angle);
+
+        // trying to solve (y2 - y1) = m (x2 - x1) + c here
+        // starting from origin point/screen center (x1, y1), end at one of the screen bounding
+        // (y2 - y1) is the equivalent of the y position
+        // (x2 - x1) is the equivalent of the x position
+        // c is the line's y-intercept, but line pass through origin, so this will be 0
+        // finalize to y = mx, or x = y/m
+
+        if (centerRelativePosition.x > 0) {
+            centerRelativePosition = new Vec3(maxX, maxX * m, 0);
+        } else {
+            centerRelativePosition = new Vec3(minX, minX * m, 0);
+        }
+
+        if (centerRelativePosition.y > maxY) {
+            centerRelativePosition = new Vec3(maxY / m, maxY, 0);
+        } else if (centerRelativePosition.y < minY) {
+            centerRelativePosition = new Vec3(minY / m, minY, 0);
+        }
+
+        // bring the position back to normal screen space (top left origin point)
+        return new Vec2(
+                (float) (centerRelativePosition.x + centerPoint.x), (float) (centerRelativePosition.y + centerPoint.y));
+    }
+
+    private boolean isInBound(Position position, Window window) {
+        return position.x() > 0
+                && position.x() < window.getGuiScaledWidth()
+                && position.y() > 0
+                && position.y() < window.getGuiScaledHeight()
+                && position.z() < 1;
+    }
+
+    private double getMarkerVisibility(Location location, ResolvedMarkerOptions resolvedMarkerOptions) {
```

Comment:
This feels like it should be part of the mapdata services, somewhere at least, rather than a Feature.

Reply:
I remember discussions about this method's implementation, but I have no idea if it's location was ever discussed. As of right now, I can't think of anything else rendering map features in-world, so this method might very well not be needed anywhere else.

Reply:
Do we not do something very similar when calculating visibility for icons on the map? Where is that done? In the MapScreen code? If so, I think it too should move.

This is not so much about code usability as encoding the "knowledge" on how to convert the attributes into a visibility percentage. It's an integral part on how the `fade` attribute is supposed to work.

Reply:
The map calculation code is in `MapDataService#calculateVisibility`. I'll move this code now too, and perhaps rename both of them to better reflect in which "space" they are used in.